src/wp-admin/includes/image.php

```diff
@@ -278,70 +282,98 @@ function wp_create_image_subsizes( $file, $attachment_id ) {
+		foreach ( $output_mime_types as $mime_index => $output_mime_type ) {
 			$editor = wp_get_image_editor( $file );
 
 			if ( is_wp_error( $editor ) ) {
 				// This image cannot be edited.
 				return $image_meta;
 			}
 
-			// Resize the image.
-			$resized = $editor->resize( $threshold, $threshold );
-			$rotated = null;
-
-			// If there is EXIF data, rotate according to EXIF Orientation.
-			if ( ! is_wp_error( $resized ) && is_array( $exif_meta ) ) {
-				$resized = $editor->maybe_exif_rotate();
-				$rotated = $resized;
-			}
-
-			if ( ! is_wp_error( $resized ) ) {
-				// Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
-				// This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
-				$saved = $editor->save( $editor->generate_filename( 'scaled' ) );
-
-				if ( ! is_wp_error( $saved ) ) {
-					$image_meta = _wp_image_meta_replace_original( $saved, $file, $image_meta, $attachment_id );
+			// If the original image's dimensions are over the threshold,
+			// scale the image and use it as the "full" size.
+			if ( $threshold && ( $image_meta['width'] > $threshold || $image_meta['height'] > $threshold ) ) {
+				// Resize the image.
+				$resized = $editor->resize( $threshold, $threshold );
+				$rotated = null;
+
+				// If there is EXIF data, rotate according to EXIF Orientation.
+				if ( ! is_wp_error( $resized ) && is_array( $exif_meta ) ) {
+					$resized = $editor->maybe_exif_rotate();
+					$rotated = $resized;
+				}
 
-					// If the image was rotated update the stored EXIF data.
-					if ( true === $rotated && ! empty( $image_meta['image_meta']['orientation'] ) ) {
-						$image_meta['image_meta']['orientation'] = 1;
+				if ( ! is_wp_error( $resized ) ) {
+					// For the default mime type scaled image is named "my_image-scaled.jpg".
+					if ( 0 === $mime_index ) {
+						// Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
+						// This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
+						$saved = $editor->save( $editor->generate_filename( 'scaled' ) );
+
+						if ( ! is_wp_error( $saved ) && ! empty( $saved ) ) {
+							$image_meta                                 = _wp_image_meta_replace_original( $saved, $file, $image_meta, $attachment_id );
+							$image_meta['sources'][ $output_mime_type ] = _wp_get_sources_from_meta( $saved );
+
+							// If the image was rotated update the stored EXIF data.
+							if ( true === $rotated && ! empty( $image_meta['image_meta']['orientation'] ) ) {
+								$image_meta['image_meta']['orientation'] = 1;
+							}
+						} else {
+							// TODO: Log errors.
+						}
+					} else {
+						// Additional mime types are named "my_image-{mime-extension}-scaled"
+						$saved = $editor->save( $editor->generate_filename( str_replace( 'image/', '', $output_mime_type ) . '-scaled' ), $output_mime_type );
```

Comment:
This doesn't look right. We shouldn't need to put the MIME type into the file name itself, instead we should use the proper extension for the MIME type. For example, the WebP version of a `my-image-scaled.jpg` should be `my-image-scaled.webp`, not `my-image-webp-scaled.jpg`.

We can get the correct `$extension` by looking it up via `array_flip( wp_get_mime_types() )` and then do the following:
```suggestion
						$saved = $editor->save( $editor->generate_filename( 'scaled', null, $extension ), $output_mime_type );
```

Reply:
> This doesn't look right. We shouldn't need to put the MIME type into the file name itself

yea, I agree and tried that originally...

> We can get the correct $extension by looking it up via array_flip( wp_get_mime_types() ) and then do the following:

I'll give this a try, I think your extension suggestion will fix the issue I saw where the image wasn't generated.

Reply:
This worked great, thanks!